modules/common/src/main/scala/surge/streams/KafkaEventSource.scala

```diff
@@ -35,34 +39,44 @@ trait EventSourceDeserialization[Event] {
 
   def shouldParseMessage(key: String, headers: Map[String, Array[Byte]]): Boolean = true
 
-  protected def onDeserializationFailure(key: String, value: Array[Byte], exception: Throwable): Unit = {
-    log.error("Unable to read event from byte array", exception)
+  private def deserializeMessage[Meta](
+      eventPlusOffset: EventPlusStreamMeta[String, Array[Byte], Meta]): Either[Meta, EventPlusStreamMeta[String, Event, Meta]] = {
+    import surge.internal.tracing.TracingHelper._
+    val span = eventPlusOffset.span
+    span.log("deserializing", Map("message size (bytes)" -> eventPlusOffset.messageBody.size.toString))
+    val payloadByteArray: Array[Byte] = eventPlusOffset.messageBody
+    val key = eventPlusOffset.messageKey
+    Try(eventDeserializationTimer.time(formatting.readEvent(payloadByteArray))) match {
+      case Failure(exception) =>
+        log.error("Unable to read event from byte array", exception)
+        span.log("failed to deserialize")
+        span.error(exception)
+        span.end()
+        Left(eventPlusOffset.streamMeta)
+      case Success(event) =>
+        span.log("deserialized", Map("event" -> event.getClass.getSimpleName))
```

Comment:
Where does this span ultimately end in the success case here? Is that still something to be added at the offset commit stage since it looks like this span is passed from the subscription to Kafka? Additionally do we want to end the span in the error case here? Or have the offset commit stage at the very end of the stream end the span on either success or failure?

Reply:
Yes, so I see three logical things here:
1. Fetching the message from Kafka
2. Deserializing the message and running the business logic
3. Offset commit stage

(1) We won't be able to capture in a span because it's deep in the Alpakka Kafka source so let's ignore it for now.
(2) It's what we capture now and the span ends upon completion of the execution of business logic
(3) I propose that the offset commit stage has its own span that "follows through" from (2).

Reply:
Cool - that makes sense to me!